File: day24/part2.py
#!/usr/bin/env python3
#
# Z3 constaint programming solution for Part 2.

import z3

# Parse input file into lists of numbers, 6 for each stone
# e.g., 19, 13, 30 @ -2,  1, -2
stones = []
for line in open('input.txt'):
    words = line.split()
    words = words[:3] + words[4:]
    words = [w.strip().replace(',', '') for w in words]
    nums = [int(w) for w in words]
    stones.append(nums)
print(len(stones), 'stones read')

# Create a solver
print('Creating solver')
solver = z3.Solver()

# Create six variables, for the position and velocity of the new stone
x = z3.BitVec('x', 64)
y = z3.BitVec('y', 64)
z = z3.BitVec('z', 64)
vx = z3.BitVec('vx', 64)
vy = z3.BitVec('vy', 64)
vz = z3.BitVec('vz', 64)

# Add constraints
# - The position of a stone is always its initial position plus its velocity
#   times time, i.e., x + vx * t == ax + vax * t    
# - For each existing stone, at some time t, the position of the stone must be
#   the same as that of the new stone
# - The time t must be positive
i = 0
for s in stones: #[:10]:

    # Unpack the variables for this stone
    x1, y1, z1, vx1, vy1, vz1 = s

    # Create a time variable for each stone, since each stone will
    # intersect the trajectory of the new stone at a different time
    i += 1
    t = z3.BitVec('t%d' % i, 64)

    # Add constraints, that the current stone must have the same position as
    # the new stone at some time t, one constraint per dimension
    solver.add(t >= 0)  # time must be positive
    solver.add(x + vx * t == x1 + vx1 * t)
    solver.add(y + vy * t == y1 + vy1 * t)
    solver.add(z + vz * t == z1 + vz1 * t)

# Check the model, find the solution
print('Solving')
assert solver.check() == z3.sat

# Get the model inputs and show answer
model = solver.model()
x = model.eval(x).as_long()
y = model.eval(y).as_long()
z = model.eval(z).as_long()
print('x/y/z:', x, y, z)
print('Part 2:', x + y + z)

File: day24/part2.go
// Find parameters of stone X (6 vals), such that pos(X) == pos(A)
// at some time, for every stone A.
//
// Optimization/constraint formulation:
// - x,y,z and vx,vy,vz for new stone are variables
// - t values for each stone are variables (not used)
// - constraints are
//       x1 + vx1 * t1 == xX + vxX * t1
//       y1 + vy1 * t1 == yX + vyX * t1
//       z1 + vz1 * t1 == zX + vzX * t1
// - Not really an objective function, except to make
//   sure there is a t for every stone

// Use Centipede constraint solver to solve this

package main

import (
	"context"
	"fmt"
	"math"

	"github.com/gnboorse/centipede"
)

const MaxInt int = 100 // need much higher for input

func part2a() {

	// Create variables x,y,z and vx,vy,vz for the position and
	// velocity of the new stone
	varNames := []centipede.VariableName{"x", "y", "z", "vx", "vy", "vz"}
	vars := centipede.Variables[int]{}
	for _, vname := range varNames {
		v := centipede.NewVariable(vname, centipede.IntRange(1, MaxInt))
		vars = append(vars, v)
	}

	// Create constraint for each stone, such that its position matches
	// that of the new stone, at some time in the future, e.g.,
	//       x1 + vx1 * t1 == xX + vxX * t1
	//       y1 + vy1 * t1 == yX + vyX * t1
	//       z1 + vz1 * t1 == zX + vzX * t1
	constraints := centipede.Constraints[int]{}
	for i, A := range stones {

		// Add a time step variable for this stone
		tVar := centipede.VariableName(fmt.Sprintf("t_%d", i))
		v := centipede.NewVariable(tVar, centipede.IntRange(1, MaxInt))
		vars = append(vars, v)

		// Add the constraint for this stone, that its position
		// is the same as that of the new stone, at some time in
		// the future
		c := centipede.Constraint[int]{Vars: varNames,
			ConstraintFunction: func(variables *centipede.Variables[int]) bool {

				// Return true if any of the variables are empty (why?)
				if variables.Find(tVar).Empty {
					return true
				}
				for _, vn := range varNames {
					if variables.Find(vn).Empty {
						return true
					}
				}

				// Get values of variables
				t := float64(variables.Find(tVar).Value)
				x := float64(variables.Find("x").Value)
				y := float64(variables.Find("y").Value)
				z := float64(variables.Find("z").Value)
				vx := float64(variables.Find("vx").Value)
				vy := float64(variables.Find("vy").Value)
				vz := float64(variables.Find("vz").Value)

				// Position this first stone at this time
				xA := A.x + t*A.vx
				yA := A.y + t*A.vy
				zA := A.z + t*A.vz

				// Position of new stone at this time
				xX := x + t*vx
				yX := y + t*vy
				zX := z + t*vz

				// Check if they match
				return close(xA, xX) && close(yA, yX) && close(zA, zX)
			},
		}

		// Add constraint to the model
		constraints = append(constraints, c)
	}

	// Solve the problem
	fmt.Println(len(vars), "variables, ", len(constraints), "constraints")
	solver := centipede.NewBackTrackingCSPSolver(vars, constraints)
	ctx := context.Background()
	success, _ := solver.Solve(ctx)

	// output results and time elapsed
	if success {
		fmt.Println("Found solution:")
		for _, variable := range solver.State.Vars {
			fmt.Printf("  %v = %v\n", variable.Name, variable.Value)
		}
	} else {
		fmt.Println("Could not find solution")
	}
}

// Determine two numbers are very close
func close(a, b float64) bool {
	return math.Abs(a-b) < .0001
}
